CaseHelper

kaapstorm opened the pull request:
## Technical Summary

I have been looking for a simple way of doing simple operations on cases. `CaseFactory` was the best option I could find, but it is not appropriate for code outside of tests.

So I wrote `CaseHelper`. Copy-pasting from its docstring:

> Initialize `CaseHelper` with an existing case:
> 
> ```python
> helper = CaseHelper(case)
> ```
> 
> Or create a case using data in the format returned by `CommCareCase.to_api_json(lite=True)`:
>
> ```python
> helper = CaseHelper()
> helper.create_case({
>     'case_id': uuid4().hex,
>     'domain': 'hawaiian-mythology',
>     'properties': {
>         'case_name': 'Namaka',
>         'case_type': 'child',
>     },
> })
> ```
> 
> Update the case:
> 
> ```python
> helper.update(properties={'description': 'Goddess of the sea'})
> ```
> 
> Add an index:
> 
> ```python
> helper.update(indices={'mother': {
>     'case_id': mother.case_id,
>     'case_type': mother.type,
>     'relationship': 'child',
> }})
> ```
> 
> Close the case:
> 
> ```python
> helper.close()
> ```
> 
> `create_case()`, `update()`, and `close()` methods all accept `user_id`, `device_id` and `xmlns` parameters to identify who/what called the operation (with reasonable defaults if you don't set them). Under the hood, all operations use `submit_case_blocks()` to submit forms.


## Safety Assurance

### Safety story

This code is not yet used.

### Automated test coverage

Hopefully test coverage is thorough. Please let me know if you can spot anything I missed.

### QA Plan

`CaseHelper` won't be QAed directly, but I would like to use it in the feature I'm working on, which will be QAed.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
CaseHelper

kaapstorm committed:
Fix assertion

snopoke reviewed (commented):
Can you describe the issue with using `CaseFactory` outside of test code?

I do like the simplicity of this API but a few points:
* Requiring a full case seems unnecessary, I think all you need is `case_id` and `domain`. 
* Specifying `domain` in the constructor is a good safety measure
* You could use the case API v0.6 code to validate the structure and create the cases. See `corehq.apps.hqcase.api.updates.handle_case_update`. I think that would make this a simple wrapper on that API which seems ideal and removes a lot of code duplication.

millerdev commented:
"Use the case API v0.6 code" to "make this a simple wrapper on that API" - seems like a great idea.

kaapstorm committed:
Wrap `handle_case_update()` instead

kaapstorm commented:
> Can you describe the issue with using `CaseFactory` outside of test code?

`CaseFactory` is only used in tests and management commands, except for `corehq/apps/zapier/views.py`, and once in `corehq/ex-submodules/casexml/apps/phone/utils.py`. So I hesitantly thought it should be OK to use it in the new Attendance Tracking feature. But then I asked a couple of other devs, and they also thought it would be best to avoid it.

> I think all you need is `case_id` and `domain`.

`CaseHelper` now accepts `case` or `case_id`, and needs `domain` to be specified.

> You could use the case API v0.6 ...

I have switched from using the format returned by `CommCareCase.to_api_json()` and validating the schema, to using the format returned by Case API v0.6's `serialize_case()` and using `handle_case_update()` for all the heavy lifting, including validation. I have also added a few more tests.

millerdev commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -1,293 +1,188 @@
+
+        @dataclass
+        class BriefCase:
+            """Like a CommCareCase, but briefer."""
```
:briefcase:

kaapstorm replied:
(Yeah, I have got to admit, that class name is my favorite part. :joy: )

millerdev commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -0,0 +1,188 @@
+
+        helper = CaseHelper(domain=domain)
+        helper.create_case({
+            'domain': 'hawaiian-mythology',
```
Is this one necessary since `domain` was passed to the constructor?

Related, but not quite the same, would it raise an error if the domains were different?

kaapstorm replied:
:thinking: Maybe I should drop that, to simplify the example.

It's not necessary. In fact, `_copy_case_data()` will drop it.

millerdev commented on corehq/apps/hqcase/case_helper.py:
```diff
@@ -0,0 +1,188 @@
+            'server_last_modified',
+            'indexed_on',
+        }
+        return {k: v for k, v in case_data.items() if k not in invalid_fields}
```
I think this method should raise an error rather than ignoring invalid items because it would surface errors sooner (rather than allowing them to pass silently).

kaapstorm replied:
If we drop this method, `handle_case_update()` will raise the validation errors we would want.

I want `CaseHelper.create_case()` to be able to accept the output of `serialize_case()`, without the user needing to know how to clean up the data first.

I am looking for a standard format for a case. Initially I thought the output of `CommCareCase.to_api_json()` could be that format, but the Case API v0.6 uses a different format. So I'm hoping that the output of `serialize_case()` can be that format, and we can pass it straight to `CaseHelper.create_case()` without a problem.

millerdev replied:
> If we drop this method, `handle_case_update()` will raise the validation errors we would want.

This sounds like a good simplification.

> I want `CaseHelper.create_case()` to be able to accept the output of `serialize_case()`, without the user needing to know how to clean up the data first.

This also sounds convenient.

> I am looking for a standard format for a case...

Sounds like there is an unsolved problem here yet. I'm hoping this was mostly you thinking out loud because I'm not sure what to suggest without digging deeper into the code.

kaapstorm replied:
> This sounds like a good simplification. ... This also sounds convenient.

I think we can have both, if we raise validation errors by default, and require the user to ask explicitly for the data to be cleaned to avoid the validation errors: dfc4cfba976

Updated tests show that validation errors are raised unless the user passes `is_serialized=True` to avoid them.

kaapstorm committed:
Fix user_id on close()

kaapstorm committed:
Raise validation errors unless is_serialized

millerdev commented on corehq/apps/hqcase/tests/test_case_helper.py:
```diff
@@ -26,11 +27,19 @@ class CaseHelperTests(TestCase):
+            case_data = serialize_case(case)
+            helper = CaseHelper(domain=DOMAIN)
+            with self.assertRaises(UserError):
+                helper.create_case(case_data)
```
Would it be difficult to make a "generator test" (with [`@generate_cases`](https://github.com/dimagi/commcare-hq/blob/14f30881c46933330194434f5d730eef95bd8e4e/corehq/util/test_utils.py#L403)) that verifies that each of the keys in
```py
        invalid_fields = {
            'case_id',
            'domain',
            'closed',
            'date_closed',
            'date_opened',
            'last_modified',
            'modified_by',
            'server_last_modified',
            'indexed_on',
        }
```
raises an error when it is the only illegal key in `case_data`?

I'm happy to attempt to clarify if this request is not clear.

kaapstorm replied:
That was a nice way of asking whether I was sure that I had all of the invalid fields :grin: 
e67463f2aa1

millerdev reviewed (approved):
Looking good.

kaapstorm committed:
Verify invalid_fields

kaapstorm committed:
Don't modify TestCase

millerdev commented on corehq/apps/hqcase/tests/test_case_helper.py:
```diff
@@ -293,6 +294,26 @@ def test_get_user_duck_username(self):
             self.assertEqual(user.username, commcare_user.username)
 
 
+@generate_cases([(f.name,) for f in CommCareCase._meta.fields], cls=TestCase)
```
Is `invalid_fields` a subset of `CommCareCase._meta.fields`? Are there any invalid fields that are not in `CommCareCase._meta.fields`? Things like

https://github.com/dimagi/commcare-hq/blob/41773d5ec28a9e05189f493d5d35be95607fac63/corehq/apps/case_search/const.py#L33-L38

kaapstorm replied:
Anything that isn't in [`BaseJsonCaseChange._properties_by_key`](https://github.com/dimagi/commcare-hq/blob/834ce5325108373b03eed3a2b46bdfb429affe08/corehq/apps/hqcase/api/updates.py#L83) is an invalid field.

:thinking: I'm doing this the wrong way round, aren't I?

...

fae79da2d2a

kaapstorm committed:
Check valid fields, instead of invalid fields

kaapstorm commented on corehq/apps/hqcase/tests/test_case_helper.py:
```diff
@@ -294,7 +294,14 @@ def test_get_user_duck_username(self):
+    [(f.name,) for f in CommCareCase._meta.fields],
+    cls=TestInvalidFields,
+)
 def test_invalid_fields(self, field_name):
```
Nice!

snopoke reviewed (commented):
@kaapstorm thanks for the updates. This is great.
